### ISSUE TITLE
Inconsistency in type checking between `function` and `const`

### ISSUE THREAD

**Issue by aardvarkk** [microsoft/TypeScript#49658]
# Bug Report

### 🔎 Search Terms

function const type checking any

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Type System Behaviour.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.2#code/GYVwdgxgLglg9mABARxADwBQEoBciCGYAnogN4BQiiATgKZQjVIBEAjAEwDMzA3OQL7lyoSLASJgcONjwBnKNRhgA5mUqIYwRBgCy+KAAsAdNUIATOAFtsWNVSp0GTFOmx8q-RLQA2s2nftHRiQOTndEQUFyCAR5RAAjfGo8bEQAXgA+RHlFFXTtW0yAzW09QxNzKxsAh3pgl0wscM8fPxqaOudQ5oEeIA)

### 💻 Code

```ts
function qux(): any {
  return "123";
}

function foo(): string {
  if (Math.random()) {
    return qux();
  } else {
    return 123; // Correctly shows TS error returning a number instead of a string
  }
}

const bar: () => string = () => {
  if (Math.random()) {
    return qux();
  } else {
    return 123; // No TS error despite returning a number instead of a string
  }
};
```

### 🙁 Actual behavior

I would expect both `foo` and `bar` to show the same TS error. In the case of `foo`, I have declared the return type as `string`, but there's a branch that returns `123` (a number). The TS compiler correctly complains about `foo`.

`bar` is declared to be a function that returns a string. However, TS for some reason doesn't complain about the same return value of `123` in `bar`.

### 🙂 Expected behavior

I believe both should trigger the same error about a bad return type.

**Comment by aardvarkk** [comment]
It looks like this is my fault for not typing the function in the assignment. I do get the expected error if I say:

`const bar: () => string = (): string => { ... `

**Comment by fatcerberus** [comment]
The assignment to `bar` should still be an error.  Do you not have `strictFunctionTypes` enabled?

**Comment by aardvarkk** [comment]
@fatcerberus Thanks for the response. I definitely try to avoid all use of `any`, but in this case it was third party code. I am making it look like my own code in order to make a small reproducible example. Anyway, I was surprised that I wasn't being given an error by TypeScript in the one circumstance but was in the other. But I think here it was just by anonymous function assignment that made it all fail.

**Comment by fatcerberus** [comment]
In your example the assignment to `bar` succeeds because the anonymous function’s return type is inferred as `any | number`, which reduces to just `any`.  That function type in turn is assignable to `() => string` due to the `any` return type.  So there’s no point where the compiler could go “oops, you’re not supposed to return a number here”.

It wouldn’t be so bad if it weren’t for the `any` getting involved - if the aforementioned inference produced, e.g. `() => string | number` for the function expression then `return 123` still wouldn’t be an error but you’d at least get an error on the assignment to `bar`.

In general you will only get type errors on individual `return` statements in functions with explicit return type annotations.
